ociclient/cache/filesystem.go

```diff
@@ -0,0 +1,348 @@
+	Size int64
+
+	index Index
+	// currentSize is the current size of the filesystem.
+	currentSize   int64
+	resetStopChan chan struct{}
+
+	// optional metrics
+	itemsCountMetric prometheus.Gauge
+	diskUsageMetric  prometheus.Gauge
+	hitsCountMetric  prometheus.Counter
+}
+
+// NewCacheFilesystem creates a new FileSystem cache.
+// It acts as a replacement for a vfs filesystem that restricts the size of the filesystem.
+// The garbage collection is triggered when a file is created.
+// When the filesystem is not used anymore fs.Close() should be called to gracefully free resources.
+func NewCacheFilesystem(identifier string, log logr.Logger, fs vfs.FileSystem, size string) (*FileSystem, error) {
+	cFs := &FileSystem{
+		Identifier: identifier,
+		log:        log,
+		FileSystem: fs,
+		index:      NewIndex(),
+	}
+	if len(size) != 0 {
+		quantity, err := resource.ParseQuantity(size)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse size %q: %w", size, err)
+		}
+		sizeInBytes, ok := quantity.AsInt64()
+		if ok {
+			cFs.Size = sizeInBytes
+		}
+	}
+
+	// load all cached files from the filesystem
+	files, err := vfs.ReadDir(fs, "/")
+	if err != nil {
+		return nil, fmt.Errorf("unable to read current cached files: %w", err)
+	}
+	for _, file := range files {
+		cFs.currentSize = cFs.currentSize + file.Size()
+		cFs.index.Add(file.Name(), file.Size(), file.ModTime())
+	}
+
+	// start hit reset counter
+	cFs.StartResetInterval()
+
+	return cFs, nil
+}
+
+// WithMetrics adds prometheus metrics to the filesystem
+// that are set by the filesystem.
+func (fs *FileSystem) WithMetrics(itemsCount, usage prometheus.Gauge, hits prometheus.Counter) {
+	fs.diskUsageMetric = usage
+	fs.hitsCountMetric = hits
+	fs.itemsCountMetric = itemsCount
+
+	if fs.diskUsageMetric != nil {
+		fs.diskUsageMetric.Set(float64(fs.CurrentSize()))
+	}
+	if fs.itemsCountMetric != nil {
+		fs.itemsCountMetric.Set(float64(fs.index.Len()))
+	}
+}
+
+// Close implements the io.Closer interface.
+// It should be called when the cache is not used anymore.
+func (fs *FileSystem) Close() error {
+	if fs.resetStopChan == nil {
+		return nil
+	}
+	fs.resetStopChan <- struct{}{}
+	return nil
+}
+
+var _ io.Closer = &FileSystem{}
+
+// StartResetInterval starts the reset counter for the cache hits.
+func (fs *FileSystem) StartResetInterval() {
+	interval := time.NewTicker(ResetInterval)
+	fs.resetStopChan = make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-interval.C:
+				fs.index.Reset()
+			case <-fs.resetStopChan:
+				interval.Stop()
+				return
+			}
+		}
+	}()
+}
+
+func (fs *FileSystem) Create(path string, size int64) (vfs.File, error) {
+	fs.writeMux.Lock()
+	defer fs.writeMux.Unlock()
+	file, err := fs.FileSystem.Create(path)
+	if err != nil {
+		return nil, err
+	}
+	fs.setCurrentSize(fs.currentSize + size)
+	fs.index.Add(path, size, time.Now())
+	if fs.itemsCountMetric != nil {
+		fs.itemsCountMetric.Inc()
+	}
+	go fs.RunGarbageCollection()
```

Comment:
yes you're right.
I also thought of a combined score but it would make the gc even more complex.
And another issue was that I currently have now real number of high the `hits` could to calculate a fixed combined.
What we could do is something like given the oldest and the most hits calculate a score upon that?

Reply:
that might work. Another option I see, could be a exclude a just added item from the next gc run.

Reply:
I think the exclusion of this one item would be harder to implement as doing a proper prioritisation. 
So now I use a value that is calculated using the the hits and the creation date.
Whereas the hits is valued more (currently 60%).